kubeflow/notebooks/prototypes/README.md

```diff
@@ -0,0 +1,131 @@
+## Goals
+
+- provide a k8 native mechanism to spawning jupyter notebooks for users
+
+- spawn a notebook within a RBAC protected namespace for a user
+
+- set a serviceAccountName within the spawning pod similar to jupyter-notebook
+
+- use k8 native authentication for the user within the protected namespace
+
+- allow similar params as kubeflow/core/prototypes/jupyterhub.jsonnet, including PVCs
+
+
+## Design
+
+The Notebooks component is an alternative to jupyterhub. The component defines a Notebook CRD and provides a notebook-controller (based on metacontroller's [CompositeController](https://metacontroller.app/api/compositecontroller/)). The Notebook CRD schema is similar to a kubernetes [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#creating-a-deployment) where it defines a [Pod Template](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#pod-template) that is used by the notebook-controller to create a Pod. An example Notebook yaml is shown below:
+
+```yaml
+apiVersion: kubeflow.org/v1alpha1
+kind: Notebook
+metadata:
+  name: training
+  namespace: resnet50
+spec:
+  template:
+    spec:
+      containers:
+      - name: notebook
+        image: gcr.io/kubeflow-images-public/tensorflow-1.10.1-notebook-cpu:v0.3.0
+        resources:
+          requests:
+            cpu: 500m
+            memory: 1Gi
+        workingDir: "/home/jovyan"
+      ttlSecondsAfterFinished: 300
+      securityContext: 
+      - fsGroup: 100
+        runAsUser: 1000
+```
+
+### User Interaction
+
+The user submits a Notebook yaml either through a UI or CLI (eg `kubectl apply -f notebook.yaml`) and the Notebook yaml is handled by the notebook-controller. The controller will create a Service and Pod within the user's namespace. The Service uses ambassador to create a reverse proxy that will route subsequent browser requests to the Pod. An example Service is shown below:
```

Comment:
yes, the controller could be changed to do something along these lines. Since our  common API Gateway is ambassador it replaces the (redundant) configurable-http-proxy in jupyterhub. In general we've wanted to move routing, authn up to the API gateway from these types of services.

Reply:
I suspect @jlzhao27's question was whether using Ambassador makes the CRD overly Kubeflow specific and whether we should generalize that by somehow letting customers use ingress.

I think one of the motivations for the CRD is to hide the creation of additional resources (e.g. the routes) needed to setup the notebook. So for this reason I think it makes sense to add the Ambasador annotation to the services.

If a user isn't using Ambassador then the annotations will simply be null op.

FWIW, I think we are moving in the direction of using ISTIO to dynamically manage routes. One of the nice things about this CRD is that we should be able to hide that change from other services (e.g. the notebook web app).

Once we switch to ISTIO all services would be behind the ISTIO gateway. So ideally the only thing that admins would have to figure out how to expose is the ISTIO gateway which could be done a number of ways (e.g. VPN or IAP).

Reply:
Cool this makes sense, will istio be relying on ingresses or just raw service annotations?

Main reason I'm advocating for ingress as well is the current set of controllers we run internally both operate on ingress annotations instead of service. Since the semantic of k8s services is cluster internal communication where as ingress is cluster external. 

 ([nginx-ingress](https://github.com/kubernetes/ingress-nginx), [heptio/contour](https://github.com/heptio/contour)) both

Reply:
i'm not sure what istio will rely on. It's possible after the switch we would need to replace the service with an ingress - or make it configurable.